adapters/Vungle/VungleAdapter/Bidding/GADMAdapterVungleBiddingRouter.m

```diff
@@ -0,0 +1,286 @@
+
+- (void)vungleAdPlayabilityUpdate:(BOOL)isAdPlayable
+                      placementID:(nullable NSString *)placementID
+                         adMarkup:(nullable NSString *)adMarkup
+                            error:(nullable NSError *)error {
+  id<GADMAdapterVungleDelegate> delegate = [self getDelegateForAdMarkup:adMarkup];
+  if (!delegate) {
+    return;
+  }
+
+  if (isAdPlayable) {
+    [delegate adAvailable];
+    return;
+  }
+
+  // Vungle SDK calls this method with isAdPlayable NO just after an ad is presented. These events
+  // should be ignored as they aren't related to a load call. Assume an ad is presented if Vungle
+  // SDK has an ad cached for this placement.
+  if ([[VungleSDK sharedSDK] isAdCachedForPlacementID:placementID
+                                             adMarkup:[delegate bidResponse]]) {
+    return;
+  }
+
+  // Vungle SDK calls this method for auto-cached placements after playing the ad.
+  // If the next placement fails to download the ad, then the SDK would call this method
+  // with isAdPlayable NO. If the delegate is already loaded, do not remove it from the
+  // tracker because the call is for the next ad.
+  if ([delegate isAdLoaded]) {
+    return;
+  }
+
+  // Ad not playable. Return an error.
+  if (error) {
+    NSLog(@"Vungle Ad Playability returned an error: %@", error.localizedDescription);
+  } else {
+    error = GADMAdapterVungleErrorWithCodeAndDescription(
+        GADMAdapterVungleErrorAdNotPlayable,
+        [NSString stringWithFormat:@"Ad is not available for placementID: %@.", placementID]);
+  }
+  [delegate adNotAvailable:error];
+  [self removeDelegate:delegate];
+}
+
+- (void)vungleWillShowAdForPlacementID:(nullable NSString *)placementID
+                              adMarkup:(nullable NSString *)adMarkup {
+  id<GADMAdapterVungleDelegate> delegate = [self getDelegateForAdMarkup:adMarkup];
+  [delegate willShowAd];
+}
+
+- (void)vungleDidShowAdForPlacementID:(nullable NSString *)placementID
+                             adMarkup:(nullable NSString *)adMarkup {
+  NSLog(@"Vungle: Did show Ad for placement ID: %@", placementID);
+  NSLog(@"Vungle: Did show Ad for ad markup: %@", adMarkup);
+}
+
+- (void)vungleAdViewedForPlacementID:(nullable NSString *)placementID
+                            adMarkup:(nullable NSString *)adMarkup {
+  id<GADMAdapterVungleDelegate> delegate = [self getDelegateForAdMarkup:adMarkup];
+  [delegate didViewAd];
+}
+
+- (void)vungleWillCloseAdForPlacementID:(nullable NSString *)placementID
+                               adMarkup:(nullable NSString *)adMarkup {
+  id<GADMAdapterVungleDelegate> delegate = [self getDelegateForAdMarkup:adMarkup];
+  [delegate willCloseAd];
+}
+
+- (void)vungleDidCloseAdForPlacementID:(nullable NSString *)placementID
+                              adMarkup:(nullable NSString *)adMarkup {
+  id<GADMAdapterVungleDelegate> delegate = [self getDelegateForAdMarkup:adMarkup];
+
+  if (!delegate) {
```

Comment:
Since you are putting this check here, do you need it at the other places like L235